### ISSUE TITLE
Allowing n-dimensional state

### ISSUE THREAD

**Issue by drhagen** [drhagen/scipy_ode#12]
As we discussed in [the original thread](https://github.com/scipy/scipy/pull/6326#issuecomment-242345435): what should we do about the shape of `y`:

1. Require that `y` be a 1-dimensional array.
    * Advantages: This simplifies the internal code. I am pretty sure this is by far the most common case, but I am not sure about scalar `y`.
    * Disadvantages: Almost everything that can broadcast in Numpy and Scipy does. There is a pretty natural extension to ODE states.
2. Allow `y` to be an n-dimensional array in the convenience functions but not in the solver classes. The convenience classes will call the appropriate reshape functions in their inputs and outputs to make this transparent to the user.
    * Advantages: Keeps from complicating the hard code (the solvers), while still providing the expected interface to the casual user.
    * Disadvantages: Inconsistencies in broadcast handling. Communication between solvers and integrators requires constant reshaping. This shouldn't affect performance, but it is king of annoying.
3. Allow `y` to be an n-dimensional array in all contexts
    * Advantages: Maximally general. Maximally consistent.
    * Disadvantages: Likely a pain to code, though I have not tried. Might affect performance at the compiled level, but I don't know how important a known array dimension is to things like Cython.

**Comment by aarchiba** [comment]
Does an option 4 make sense? solvers use 1d arrays internally but convert to the user's shape on input and output?

**Comment by drhagen** [comment]
> Does an option 4 make sense? solvers use 1d arrays internally but convert to the user's shape on input and output?

So this is the interface of option 3, but as much of the internals of option 2 as can be managed?

**Comment by aarchiba** [comment]
Indeed. Or in an ideal world, it is option 2 except it pushes the conversion code into the solver base class, so solver writers don't need to deal with it either. The idea is that some users are going to be using the raw solver interface and not the convenience functions; it'd be nice if they too could benefit from flexible-shape RHSs. One example that comes to mind is linear sensitivity analysis, where one tries to get the derivatives of an ODE solution with respect to the initial conditions by integrating the Jacobian alongside the ODE itself; though since one wants the original (vector) solution as well as the (matrix) Jacobian it still requires some jiggery-pokery from the user.
